datafusion/physical-plan/src/aggregates/row_hash.rs

```diff
@@ -787,4 +752,34 @@ impl GroupedHashAggregateStream {
         self.group_ordering = GroupOrdering::Full(GroupOrderingFull::new());
         Ok(())
     }
+
+    /// returns true if there is a soft groups limit and the number of distinct
+    /// groups we have seen is over that limit
+    fn hit_soft_group_limit(&self) -> bool {
+        let Some(group_values_soft_limit) = self.group_values_soft_limit else {
+            return false;
+        };
+        group_values_soft_limit <= self.group_values.len()
+    }
+
+    // common function for signalling end of processing of the input stream
+    fn set_input_done_and_produce_output(&mut self) -> Result<()> {
```

Comment:
The original signature implies this can return early as well as may return a `RecordBatch`, neither of which is true